### ISSUE TITLE
Inconsistent gradient shape

### ISSUE THREAD

**Issue by vlukes** [sfepy/sfepy#876]
The gradient shape returned by `terms_basic.dq_grad()` is (..., dim, n_comp) while `crefcoors.evaluate_in_rc()` gives (..., n_comp, dim).

According to [1], gradient of a vector field equals the transpose of the Jacobian matrix, which is given by

J_{ij} = \partial f_i  / \partial x_j

and has the shape of (n_comp, dim). So the gradient should have the shape (dim, n_comp).

But e.g. [2] says that the gradient of a vector field is organized into a single matrix known as the Jacobian matrix. See also this [3] discussion :)

[1] https://en.wikipedia.org/wiki/Gradient
[2] https://towardsdatascience.com/step-by-step-the-math-behind-neural-networks-d002440227fb
[3] https://math.stackexchange.com/questions/156880/what-does-it-mean-to-take-the-gradient-of-a-vector-field

**Comment by rc** [comment]
Thanks for bringing this up. I would prefer, also based on the links you found, the `evaluate_at()` version, i.e. (n_comp, dim).

**Comment by vlukes** [comment]
> I would prefer, also based on the links you found, the `evaluate_at()` version, i.e. (n_comp, dim).

I do not prefer this variant, because it means that the gradient of a scalar function is (1, dim), which is inconsistent with how the gradient is understood elsewhere in the code.

`dq_grad()` is also more wired in the code than `evaluate_at()`.

**Comment by rc** [comment]
Yes, it might be more difficult. Note, however, that the current `FieldVariable.evaluate()` is internally inconsistent - the 'val' mode returns `(n_el, n_qp, n_comp, 1)`, while 'grad' returns `(n_el, n_qp, dim, n_comp)`. The 'div' and 'cauchy_strain' modes are IMO consistent with the 'val' mode.

 IIRC the origin of this "error" is connected with the FE bases, that have shape `bf.shape = (n_qp, 1, n_ep)` and `bfg.shape = (n_qp, dim, n_ep)`, to facilitate multiplication with element the DOF vector.

So, apart from the maybe better "mathematical" sense (and C layout) of `Field.evaluate_at()` in which in gu_ij = d(u_i)/d(x_j) is stored row-by-row (see also the definition of the deformation gradient), we would need to change 'val' and maybe 'cauchy_strain' modes of `FieldVariable.evaluate()` together with `Field.evaluate_at()` to be really consistent . Now changing just the 'grad' mode of `FieldVariable.evaluate()` seems easier(?) Terms would have to be updated in any case.

So I do not know - is it worth changing now?

**Comment by vlukes** [comment]
- `de_grad()`, `evaluate_at()` and also `ev_def_grad()`:  (n_comp, dim)
- `ev_grad()` (this uses `dq_grad()`):  (dim, n_comp)

**Comment by rc** [comment]
Actually `de_grad` and other multi-linear terms might fixing. First the 'el_avg' and other modes are not supported, and secondly for non-scalar output such as de_grad, the output axis order is undefined, as it depends on `set()` call in `find_free_indices()`. On my laptop, it returns the same thing as `ev_grad`. =:)  The fix would be simple: use `sorted()`. Then it would be possible to control the output order in a term definition simply by using `i.j` or `j.i`.
